### DISCUSSION TITLE
Recent hwloc breaks shell due do competition error?

### DISCUSSION THREAD

**Discussion by woj-tek** [Homebrew/discussions#5471] [Everyday usage]
### Output of `brew config`

```shell
$ brew config
HOMEBREW_VERSION: 4.3.8-20-gc524e5c
ORIGIN: https://github.com/Homebrew/brew
HEAD: c524e5c2fc68c162051ce081707e4945a280a62a
Last commit: 4 hours ago
Core tap HEAD: 8737111f976c5c9c8b2eca205fc3185ac722333c
Core tap last commit: 15 minutes ago
Core tap JSON: 02 Jul 12:05 UTC
Core cask tap HEAD: da2cffd1bbf201d2b82baa370d36a94bf1cd518c
Core cask tap last commit: 15 minutes ago
Core cask tap JSON: 02 Jul 12:05 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.3.3 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.3/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.45.2 => /opt/homebrew/bin/git
Curl: 8.6.0 => /usr/bin/curl
macOS: 14.5-arm64
CLT: 15.3.0.0.1.1708646388
Xcode: 15.4
Rosetta 2: false
```

### Output of `brew doctor`

```shell
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed casks are deprecated or disabled.
You should find replacements for the following casks:
  dozer
  meld
  skitch

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  docker-machine-completion
  python-click
  mitmproxy
  docker-compose-completion

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  neofetch
  openssl@1.1
  python-pytz
  youtube-dl

Warning: You have the following deprecated, official taps tapped:
  Homebrew/homebrew-cask-versions
  Homebrew/homebrew-cask-fonts
  Homebrew/homebrew-cask-drivers
Untap them with `brew untap`.

Warning: Putting non-prefixed coreutils in your path can cause GMP builds to fail.

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python-pytz
  python-click
  dosbox-x
  numpy
  certifi
  handbrake
  pyyaml
```

### Description of issue

After running brew upgrade today new shell opens with completion error from `hwloc`:

```
-bash: /opt/homebrew/etc/bash_completion.d/hwloc: line 210: syntax error near unexpected token `<'
-bash: /opt/homebrew/etc/bash_completion.d/hwloc: line 210: `		COMPREPLY=( "<integer>" "" )'
-bash: /opt/homebrew/etc/bash_completion.d/hwloc: line 211: syntax error near unexpected token `;;'
-bash: /opt/homebrew/etc/bash_completion.d/hwloc: line 211: `		;;'
wojtek@atlantiscity.local ~/dev $
```

From what I see formula was upgraded in last 16h (PR: https://github.com/Homebrew/homebrew-core/pull/175881 by @chenrui333 ). 

Looking at the sources of `hwloc` it completion file has more or less same syntax (especially the offending one) for a long time: https://github.com/open-mpi/hwloc/commit/14dc479399b0b3e03a7f5478560fb8c8e2dbc11b#diff-913da97e97dc7eaf7ef6ddeed0178288fb0b4e59ceecaef9c4ff208faf8eb4a7R189

I tried removing it but it fails due to it being a dependnecy:

```
$ brew remove hwloc
Error: Refusing to uninstall /opt/homebrew/Cellar/hwloc/2.11.0
because it is required by corsixth, ffmpeg@6, gstreamer, openvino, tbb and xmrig, which are currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies hwloc
```

(I was pondering commenting in the linked PR but opted out for a discussion to avoid generating noise)

**Comment by woj-tek** [comment]
Actually I managed to remove all packages that require `hwloc` and finally got rid of it… though the shell issue may still be happening.

**Comment by jonathanstallings** [comment]
I am seeing the same issue.

Running [shellcheck](https://github.com/koalaman/shellcheck) on the bash completion file for `hwloc` revealed several instances of unclosed double-quoted strings. Note the unclosed `taskset"` below, which is [from hwloc bash completion file in current master as of this writing](https://github.com/open-mpi/hwloc/blob/603b27271ee24ab5ff5bf6388f3803e440cbf152/contrib/completion/bash/hwloc#L151).
```bash
	        COMPREPLY=( "hwloc" "list" taskset" )
```

This appears to be from this commit:
https://github.com/open-mpi/hwloc/commit/e25888c9e4d1dcda8ff607f51207ca49cf89e15a

Properly closing these quoted strings resolved this issue for me, but it will need to be resolved upstream.

**Comment by sleetdrop** [comment]
I just unlink the bash completion script provided by hwloc, and wait for the maintainers to fix it in next version.
`unlink /opt/homebrew/etc/bash_completion.d/hwloc`

**Comment by jeffbyrnes** [comment]
It appears they’ve been fixed in open-mpi/hwloc#673 and open-mpi/hwloc#674.

**Reply by jeffbyrnes** [reply]
Also, Homebrew/homebrew-core#175881 added this formula to the autobump list, so we should see an updated formula for 2.11.1 soon.

**Reply by acdcfanbill** [reply]
I just did a update/upgrade and the hwloc bash completion issue is fixed for me now. Thanks everyone!
